src/databricks/labs/ucx/hive_metastore/federation.py

```diff
@@ -31,58 +48,161 @@ def enable(self):
+            raise ValueError('Hive Metastore version not found')
+        major_version = re.match(r'(\d+\.\d+)', version)
+        if not major_version:
+            raise ValueError(f'Invalid Hive Metastore version: {version}')
+        version = major_version.group(1)
+        ext_hms = replace(self._split_jdbc_url(jdbc_url), version=version)
+        supported_versions = self.supported_db_vers.get(ext_hms.db_type)
+        if not supported_versions:
+            raise ValueError(f'Unsupported Hive Metastore: {ext_hms.db_type}')
+        if version not in supported_versions:
+            raise ValueError(f'Unsupported Hive Metastore Version: {ext_hms.db_type} - {version}')
+        if not ext_hms.user:
+            ext_hms = replace(
+                ext_hms,
+                user=self._get_value_from_config_key(spark_config, 'spark.hadoop.javax.jdo.option.ConnectionUserName'),
+            )
+        if not ext_hms.password:
+            ext_hms = replace(
+                ext_hms,
+                password=self._get_value_from_config_key(
+                    spark_config, 'spark.hadoop.javax.jdo.option.ConnectionPassword'
+                ),
+            )
+        return ext_hms
+
+    @classmethod
+    def _split_jdbc_url(cls, jdbc_url: str) -> ExtHms:
+        # Define the regex pattern to match the JDBC URL components
+        pattern = re.compile(
+            r'jdbc:(?P<db_type>[a-zA-Z0-9]+)://(?P<host>[^:/]+):(?P<port>\d+)/(?P<database>[^?]+)(\?user=(?P<user>[^&]+)&password=(?P<password>[^&]+))?'
+        )
+        match = pattern.match(jdbc_url)
+        if not match:
+            raise ValueError(f'Unsupported JDBC URL: {jdbc_url}')
+
+        db_type = match.group('db_type')
+        host = match.group('host')
+        port = match.group('port')
+        database = match.group('database')
+        user = match.group('user')
+        password = match.group('password')
+
+        return ExtHms(db_type, host, port, database, user, password, None)
+
+    def _register_federated_catalog(self, connection_info) -> CatalogInfo:
         try:
-            return self._workspace_client.catalogs.create(
+            return self._ws.catalogs.create(
                 name=connection_info.name,
                 connection_name=connection_info.name,
                 options={"authorized_paths": self._get_authorized_paths()},
             )
         except BadRequest as err:
             if err.error_code == 'CATALOG_ALREADY_EXISTS':
                 logger.info(f'Catalog {connection_info.name} already exists')
-                for catalog_info in self._workspace_client.catalogs.list():
+                for catalog_info in self._ws.catalogs.list():
                     if catalog_info.name == connection_info.name:
                         return catalog_info
             raise err
 
-    def _get_or_create_connection(self, name: str) -> ConnectionInfo:
+    def _get_or_create_int_connection(self, name: str) -> ConnectionInfo:
```

Comment:
I mistakenly read "init", please avoid confusion by using the full names

```suggestion
    def _get_or_create_internal_connection(self, name: str) -> ConnectionInfo:
```

Reply:
Sure